utils_cv/detection/model.py

```diff
@@ -0,0 +1,283 @@
+
+        # and a learning rate scheduler
+        self.lr_scheduler = torch.optim.lr_scheduler.StepLR(
+            self.optimizer, step_size=3, gamma=0.1
+        )
+
+    def __getattr__(self, attr):
+        if attr in self.__dict__:
+            return self.__dict__[attr]
+        raise AttributeError(
+            "'{}' object has no attribute '{}'".format(
+                type(self).__name__, attr
+            )
+        )
+
+    def _setup_data(self, dataset: Dataset, train_ratio: float = 0.8):
+        """ create training and validation data loaders
+        """
+        self.train_ds, self.test_ds = dataset.split_train_test(
+            train_ratio=train_ratio
+        )
+
+        self.train_dl = DataLoader(
+            self.train_ds,
+            batch_size=self.batch_size,
+            shuffle=True,
+            num_workers=4,
+            collate_fn=collate_fn,
+        )
+
+        self.test_dl = DataLoader(
+            self.test_ds,
+            batch_size=self.batch_size,
+            shuffle=False,
+            num_workers=4,
+            collate_fn=collate_fn,
+        )
+
+    def fit(self, epochs: int, print_freq: int = 10) -> None:
+        """ The main training loop. """
+        self.losses = []
+        self.epochs = epochs
+        for epoch in range(self.epochs):
+            # train for one epoch, printing every 10 iterations
+            logger = train_one_epoch(
+                self.model,
+                self.optimizer,
+                self.train_dl,
+                self.device,
+                epoch,
+                print_freq=print_freq,
+            )
+            self.losses.append(logger.meters["loss"].median)
+
+            # update the learning rate
+            self.lr_scheduler.step()
+
+            # evaluate
+            self.evaluate(dl=self.train_dl)
+
+    def plot_losses(self, figsize: Tuple[int, int] = (10, 5)) -> None:
+        """ Plot training loss from fitting. """
+        fig = plt.figure(figsize=figsize)
+        ax = fig.add_subplot(111)
+        ax.set_xlim([0, self.epochs - 1])
+        ax.set_xticks(range(0, self.epochs))
+        ax.set_title("Loss over epochs")
+        ax.set_xlabel("epochs")
+        ax.set_ylabel("loss")
+        ax.plot(self.losses)
+
+    def evaluate(self, dl: DataLoader = None) -> CocoEvaluator:
+        """ eval code on validation/test set. """
+        if dl is None:
+            dl = self.test_dl
+        return evaluate(self.model, dl, device=self.device)
+
+    def get_model(self) -> nn.Module:
+        """ returns the model object. """
+        return self.model
+
+    def inference(self, im_path: Union[str, Path]):
+        """ Performs inferencing on an image path. """
+        im = Image.open(im_path)
+        transform = transforms.Compose([transforms.ToTensor()])
+        im = transform(im).cuda()
+        model = self.get_model().eval()  # eval mode
+        with torch.no_grad():
+            pred = model([im])
+        return pred
```

Comment:
Could first create a list of BBox objects, and then return that. That's so that we only ever return/input BBox objects and avoid movind around coordinates, labels, etc.